### ISSUE TITLE
Better integrate Electron>=8.1.1's spell checking

### ISSUE THREAD

**Issue by globau** [nativefier/nativefier#283]
applications built with nativefier don't perform spell checking, which is unexpected, and out of alignment with how sites work in-browser.

it would be nice if something like the `electron-spell-check-provider` package was used to enable spell checking on fields.

---

**EDIT from @ronjouch on 2019/03/11**: see https://github.com/jiahaog/nativefier/issues/283#issuecomment-597668972 , Electron ≥ 8.1.1 supports spellchecking natively, and it's (crudely) usable by simply asking for it through BrowserOptions. However, it lacks basic features (an easy-to-use flag, a context menu letting users accept suggestions and switch languages, and maybe enabling spellcheck by default), so this is what this issue is about now.

**EDIT from @ronjouch on 2020/04/22**: see https://github.com/sindresorhus/electron-context-menu/releases/tag/v1.0.0 , `electron-context-menu@1.0.0` (which I bumped Nativefier 8.0.7 to) includes _"built-in support for spellchecking (https://github.com/sindresorhus/electron-context-menu/pull/94), see https://github.com/sindresorhus/electron-context-menu/commit/71c5d2e1fb28f13fea35b1be076909f463826f55"_ . Maybe worth a look if you're interested in contributing!

**Comment by eladnava** [comment]
Would love to make this happen. 

@jiahaog Can you point me in the right direction for completing this? 👍

**Comment by ronjouch** [comment]
Ping @globau @eladnava: @Thatkookooguy tried implemented spell checking, maybe see https://github.com/jiahaog/nativefier/issues/222 if you feel like collaborate on something.

**Comment by schippie** [comment]
I would also like to voice my support for this. And was wondering if @ronjouch  you tried using something like: https://github.com/electron-userland/electron-spellchecker

It seems that this library will work on any regular inputs as it listens to the document input event. The only situation where this will cause problems would be for applications that modify the input event / block it's event propagation. And there isn't anything that can be done about that really.

**Comment by tjkohli** [comment]
+1 this is a deal breaker for us if it's not available

**Comment by basu90** [comment]
+1 it would be great

**Comment by tjkohli** [comment]
I would also like to know.

**Comment by nicoh88** [comment]
+1

**Comment by uwekoenig** [comment]
> 
> 
> +1 this is a deal breaker for us if it's not available

I would really like to see this feature.

**Comment by e1ven** [comment]
You can enable spellchecking on your own now that [this pr](https://github.com/electron/electron/pull/20692) is merged into Electron.

You'll need to pass a parameter to upgrade the version of Electron from the default, and another to enable the feature.

` --browserwindow-options '{ "webPreferences": { "spellcheck": true } }' -e 8.1.1`

**Comment by eladnava** [comment]
@e1ven Legend! Thanks for that 🎉

**Comment by eladnava** [comment]
FYI you will need to upgrade the version of `electron-packager` that `nativefier` depends on to the latest, otherwise bundling any Electron version above 5 will crash. Run this in `nativefier` directory:
```
npm install electron-packager@latest
```

Then, run `nativefier` again to regenerate the app.

**Comment by ronjouch** [comment]
Thanks @e1ven @eladnava 🙂. Regarding the need to upgrade `electron-packager`, I encourage everyone interested to give branch `ts` a try, it ships with the latest everything, including electron-packager and electron 8.1.1. I just confirmed spellchecking works in this branch with `nativefier --browserwindow-options '{ "webPreferences": { "spellcheck": true } }' 'https://www.w3schools.com/tags/tryit.asp?filename=tryhtml_textarea'` .

To those of you willing to give this branch a drive, go ahead, it's finished, currently being reviewed by Jia, and testing is more than welcome. See the build instructions at https://github.com/jiahaog/nativefier/pull/898#issuecomment-581062794 , and please report any issues with a `[ts] ` prefix.

Then, once #898 ships, help welcome to make this feature actually usable: a new flag to enable it more easily than passing `browserwindow-options`, maybe make it default, and adding context menu entries to accept corrections and switch languages.

**Comment by eladnava** [comment]
Thanks @ronjouch, gave `ts` a try, seems to work well. Only issue is #921, but that's present in `master` as well. Thanks for all your hard work on this time-saving app!

**Comment by fia5000** [comment]
spellcheck works great but how do we enable dictionary suggestions in the context menu for things spell check marks as wrong

**Comment by AndriusWild** [comment]
I just rebuilt an app with the `--browserwindow-options '{ "webPreferences": { "spellcheck": true } }'` option and seems like it did not make any difference. The electron version is 8.2.0
Do I still need to install any "latest" version in order to make the spellcheck work?

**Comment by ronjouch** [comment]
Everybody interested, see https://github.com/sindresorhus/electron-context-menu/releases/tag/v1.0.0 , `electron-context-menu@1.0.0` (which I bumped Nativefier 8.0.7 to) includes _"built-in support for spellchecking (https://github.com/sindresorhus/electron-context-menu/pull/94), see https://github.com/sindresorhus/electron-context-menu/commit/71c5d2e1fb28f13fea35b1be076909f463826f55"_ .

Worth a look for those of you interested in contributing!

**Comment by benpro** [comment]
Spellcheck is enabled on v9.0.0. But, how to change the language?

**Comment by pqnhan** [comment]
How to turn off spell checking?
Sometimes it's annoying
![image](https://user-images.githubusercontent.com/24197589/109740072-37eaa900-7bfd-11eb-862c-4937252c005b.png)

**Comment by llabusch93** [comment]
> Spellcheck is enabled on v9.0.0. But, how to change the language?

Are there any news / someone figured out how to change the spellcheck language?

**Comment by llabusch93** [comment]
Okay I found how to define that:
https://www.electronjs.org/docs/tutorial/spellchecker#how-to-set-the-languages-the-spellchecker-uses

But how can we call `BrowserWindow.session.setSpellCheckerLanguages()` to define the wanted languages?

**Comment by llabusch93** [comment]
> How to turn off spell checking?
> Sometimes it's annoying
> ![image](https://user-images.githubusercontent.com/24197589/109740072-37eaa900-7bfd-11eb-862c-4937252c005b.png)

@pqnhan you just need to run nativefier with this option: `--browserwindow-options '{ "webPreferences": { "spellcheck": false } }'`

**Comment by TheCleric** [comment]
> But how can we call BrowserWindow.session.setSpellCheckerLanguages() to define the wanted languages?

@ronjouch I have a couple ideas on this. Both involve injected JS:
1. Support an ipcMain event like `set-spell-checker-languages` that will accept an event argument from an injected JS file that will pass it on to the session function.
2. A more generic `session-interaction` event that allows calling session functions and properties dynamically, so future electron session expansions can be supported automatically.

For 2 I have a working proof of concept. I have added the following to the mainWindow.ts:

```typescript

  ipcMain.on('session-interaction', (event, request) => {
    let result = null;
    try {
      if (request.func) {
        if (request.funcArgs === undefined || request.funcArgs === null) {
          request.funcArgs = [];
        }

        if (typeof request.funcArgs[Symbol.iterator] !== 'function') {
          request.funcArgs = [request.funcArgs];
        }
        result = mainWindow.webContents.session[request.func](
          ...request.funcArgs,
        );
      } else if (request.property) {
        result = mainWindow.webContents.session[request.property];
      }
    } catch (error) {
      console.log('ERROR: session-interaction')
      console.log(request);
      console.log(error);
      result = error;
    }

    event.reply('session-interaction-reply', result);
  });
```

which now allows me to inject the following JS that will set the dictionary succesfully:

```javascript
const electron = require('electron');

electron.ipcRenderer.on('session-interaction-reply', (event, result) => {
    console.log('session-interaction-reply', event, result)
});

electron.ipcRenderer.send('session-interaction', { property: 'availableSpellCheckerLanguages', });
electron.ipcRenderer.send('session-interaction', { func: 'setSpellCheckerLanguages', funcArgs: [['fr']], });

```

**Comment by TheCleric** [comment]
Also @llabusch93 even if we can get Nativefier to support this, Electron's spellchecker comes with a pretty big caveat that I hit personally:

> On macOS as we use the native APIs there is no way to set the language that the spellchecker uses. By default on macOS the native spellchecker will automatically detect the language being used for you.

So for MacOS, the `setSpellCheckerLanguages` does nothing.

**Comment by ronjouch** [comment]
> @ronjouch I have a couple ideas on this. Both involve injected JS:
> 
> 1. Support an ipcMain event like `set-spell-checker-languages` that will accept an event argument from an injected JS file that will pass it on to the session function.
> 
> 2. A more generic `session-interaction` event that allows calling session functions and properties dynamically, so future electron session expansions can be supported automatically. And could be used for things involving the session where we don't feel like adding a CLI argument.
> 
> 
> For 2 I have a working proof of concept. I have added the following to the mainWindow.ts: [...]

@TheCleric your new plumbing looks good. Then,

1. If your intention is to be able to tell users "you can now set spellchecker language through a `--inject`ed script calling `electron.ipcRenderer.send('session-interaction', { func: 'setSpellCheckerLanguages', funcArgs: [['fr']], });` ", I'm okay to merge and document this (along with how to disable spellchecking).
2. If you want to go further, what about detecting system locale on startup and setting it automatically to `["<detectedLocale>", "en"]` ? [StackOverflow offers a few ideas](https://stackoverflow.com/questions/46072248/node-js-how-to-detect-user-language), and obviously it's super OS-specific, so it'd be reasonable to use a dep for this, [os-locale](https://github.com/sindresorhus/os-locale) looks clean, minimal, and maintained by somebody trustworthy.
4. If you want to go further further, we want an app context menu (like browsers do) letting users [1. Enable/disable spellchecking, 2. Pick locale], and persist it to json. This has the advantage to be easy and usable (because already known to users, because similar to browsers), and to not need extra flags/injects. We already have a context menu, this sounds feasible.
3. As last resort, as usual, we could have an extra flag. Something like `--spellcheck`, supporting enabling/disabling, and also supporting defining languages:
    1. `--spellcheck true`
    2. `--spellcheck false`
    3. `--spellcheck en,fr`

**Comment by TheCleric** [comment]
> @TheCleric your new plumbing looks good. Then,
> 
> 
> 
> 1. If your intention is to be able to tell users "you can now set spellchecker language through a `--inject`ed script calling `electron.ipcRenderer.send('session-interaction', { func: 'setSpellCheckerLanguages', funcArgs: [['fr']], });` ", I'm okay to merge and document this (along with how to disable spellchecking).

Yeah that seems the simplest path.

> 2. If you want to go further, what about detecting system locale on startup and setting it automatically to `["<detectedLocale>", "en"]` ? [StackOverflow offers a few ideas](https://stackoverflow.com/questions/46072248/node-js-how-to-detect-user-language), and obviously it's super OS-specific, so it'd be reasonable to use a dep for this, [os-locale](https://github.com/sindresorhus/os-locale) looks clean, minimal, and maintained by somebody trustworthy.

I believe Electron does that automatically. From their docs: `By default the spellchecker will enable the language matching the current OS locale.`

> 4. If you want to go further further, we want an app context menu (like browsers do) letting users [1. Enable/disable spellchecking, 2. Pick locale], and persist it to json. This has the advantage to be easy and usable (because already known to users, because similar to browsers), and to not need extra flags/injects. We already have a context menu, this sounds feasible.

So like on right click?

> 3. As last resort, as usual, we could have an extra flag. Something like `--spellcheck`, supporting enabling/disabling, and also supporting defining languages:
> 
>     1. `--spellcheck true`
> 
>     2. `--spellcheck false`
> 
>     3. `--spellcheck en,fr`

Honestly, don't hate this option at all.

**Comment by llabusch93** [comment]
Hey @ronjouch and @TheCleric ,

saw that your modifications are merged, cool! So is there now a way to set spellchecking languages with nativefier?

Thanks for you work!

**Comment by TheCleric** [comment]
> Hey @ronjouch and @TheCleric ,
> 
> saw that your modifications are merged, cool! So is there now a way to set spellchecking languages with nativefier?
> 
> Thanks for you work!

If you are building a non-mac app, yes. (Electron on Mac just uses whatever the current selected OS language is and can't change it due to limitations on Mac).

To set the spellcheck language on Windows and Linux, see the example on the new session-interaction api here: https://github.com/nativefier/nativefier/blob/master/docs/api.md#example

This will guide you on how to create a JS file that will set the spellcheck language that can be injected via `--inject` into the app.

**Comment by edsenabr** [comment]
@TheCleric :: I've just tested that new  session-interaction api on nativefier@43.0.0 and I'm not getting any response from the emitted event; I'm trying the sample code provided:

```
const request = {
  property: 'availableSpellCheckerLanguages',
};
electron.ipcRenderer.send('session-interaction', request);

electron.ipcRenderer.on('session-interaction-reply', (event, result) => {
    console.log('session-interaction-reply', event, result.value)
});

```
... my handler is never called

**Comment by TheCleric** [comment]
> BTW, a recursive grep for "session-interaction" on the whole nativefier code tree got me nothing. Do I need to upgrade to an experimental release on another channel ?
> 

My apologies. It doesn't appear to have made it into the npm release yet. If you pull this repo it would be in the master branch. Sorry for my confusion.

**Comment by ronjouch** [comment]
> > BTW, a recursive grep for "session-interaction" on the whole nativefier code tree got me nothing. Do I need to upgrade to an experimental release on another channel ?
> 
> It doesn't appear to have made it into the npm release yet. If you pull this repo it would be in the master branch. Sorry for my confusion.

Hi there; I just released Nativefier 43.0.1 with @TheCleric's patch in, it will be on npm in a few minutes for you to try without having to use a dev build.

**Comment by sergiomb2** [comment]
> > Hey @ronjouch and @TheCleric ,
> > saw that your modifications are merged, cool! So is there now a way to set spellchecking languages with nativefier?
> > Thanks for you work!
> 
> If you are building a non-mac app, yes. (Electron on Mac just uses whatever the current selected OS language is and can't change it due to limitations on Mac).
> 
> To set the spellcheck language on Windows and Linux, see the example on the new session-interaction api here: https://github.com/nativefier/nativefier/blob/master/docs/api.md#example
> 
> This will guide you on how to create a JS file that will set the spellcheck language that can be injected via `--inject` into the app.

now example is here https://github.com/nativefier/nativefier/blob/master/API.md#example ans works for me
